### PR TITLE
Added two new unreleased Model Racing games

### PR DESCRIPTION
**Pull request by janniz**
Those games was never released by Model Racing to the public.
The assembler source files for those games where extracted from the original floppy disks used by the former Model Racing developer Adolfo Melilli (adolfo@melilli.com).
The disks where retrieved by Alessandro Bolgia (xadhoom76@gmail.com) and Lorenzo Fongaro (lorenzo.fongaro@virgilio.it) and dumped by Piero Andreini (pieroandreini@gmail.com) using KryoFlux hardware and software.
Subsequently Jean Paul Piccato (j2pguard-spam@yahoo.com) mounted the images and compiled the source files, managed to set up a romset and wrote a mame driver that aims to reproduce in the most faithful way the work of Melilli at Model Racing in late '70s.

The game driver is not based on hardware inspection and is solely derived from assumptions I've made looking at the assembler code and comments written into the source files of the games. Several of those hypothesis came following the directions of previous yet contemporary Model Racing works (Eg: Claybuster) and where confirmed by Melilli himself.

Being unreleased these games do not have an official name, thus the names used in the source files were used instead.

In this PR there are also two fixes for the debug wave file generated by the sn76477 device.

### CODE REVIEW

**Comment by stilett0** [comment]
Quasi-related: Do you think you could enquire about this game and see if it had indeed been under development and if it had been given a title (perhaps "Pallanuoto"?)
https://mamedev.emulab.it/undumped/index.php?title=Unknown_Water_Polo_Game

Also there is a rumor that Model Racing had licensed or perhaps cloned/bootlegged Looping, any word on that?

**Comment by janniz** [comment]
> Quasi-related: Do you think you could enquire about this game and see if it had indeed been under development and if it had been given a title (perhaps "Pallanuoto"?)
> https://mamedev.emulab.it/undumped/index.php?title=Unknown_Water_Polo_Game
> 
> Also there is a rumor that Model Racing had licensed or perhaps cloned/bootlegged Looping, any word on that?

I'm asking Adolfo about your requests just now. As soon as I have some news I'll reply back to you.

File: src/devices/sound/sn76477.cpp [hunk 1]
```diff
@@ -1996,7 +1998,7 @@ void sn76477_device::sound_stream_update(sound_stream &stream, stream_sample_t *
 		 */
 		*buffer++ = (((voltage_out - OUT_LOW_CLIP_THRESHOLD) / (OUT_CENTER_LEVEL_VOLTAGE - OUT_LOW_CLIP_THRESHOLD)) - 1) * 32767;
 
-		if (LOG_WAV && LOG_WAV_ENABLED_ONLY && !m_enable)
+		if (LOG_WAV && (!LOG_WAV_ENABLED_ONLY || (LOG_WAV_ENABLED_ONLY && !m_enable)))
```
Comment (cuavas) [inline src/devices/sound/sn76477.cpp]: You could write the expression as `(LOG_WAV && (!LOG_WAV_ENABLED_ONLY || !m_enable))` and make it more obvious what it does, but why did you change the condition?  Is it out of sync with something else?

File: src/mame/audio/8080bw.cpp [hunk 2]
```diff
@@ -1397,3 +1397,250 @@ WRITE8_MEMBER( _8080bw_state::darthvdr_08_w )
...
+  m_sn->set_oneshot_params(CAP_U(10), RES_K(100));
+  m_sn->set_vco_mode(0);
+  m_sn->set_mixer_params(0, 0, 0);
+  m_sn->set_envelope_params(1, 0);
+  m_sn->set_enable(0);
+  m_sn->add_route(0, "discrete", 1.0, 0);
+
+  DISCRETE(config, m_discrete, cane_discrete);
+  m_discrete->add_route(ALL_OUTPUTS, "mono", 1.0);
+}
```
Comment (cuavas) [inline src/mame/audio/8080bw.cpp]: Please use one tab per level of scope.  This applies to other places as well, e.g. I/O ports.

File: src/mame/drivers/8080bw.cpp [hunk 3]
```diff
@@ -3560,6 +3560,443 @@ void _8080bw_state::init_invmulti()
...
+  PORT_BIT( 0x80, IP_ACTIVE_HIGH, IPT_BUTTON1 ) PORT_IMPULSE(2)
+
+  PORT_START("VR1")
+  PORT_ADJUSTER( 80, "VR1 - SFX from 76477" )
+
+  PORT_START("VR2")
+  PORT_ADJUSTER( 90, "VR2 - TOS music" )
+
+  PORT_START("VR3")
+  PORT_ADJUSTER( 70, "VR3 - Shoot SFX from 555" )
```
Comment (cuavas) [inline src/mame/drivers/8080bw.cpp]: These belong in an audio device.

File: src/mame/includes/8080bw.h [hunk 4]
```diff
@@ -177,6 +186,11 @@ class _8080bw_state : public mw8080bw_state
 	DECLARE_READ8_MEMBER(invmulti_eeprom_r);
 	DECLARE_WRITE8_MEMBER(invmulti_eeprom_w);
 	DECLARE_WRITE8_MEMBER(invmulti_bank_w);
+	DECLARE_WRITE8_MEMBER(cane_sh_port_1_w);
+	DECLARE_WRITE8_MEMBER(cane_music_w);
+	DECLARE_WRITE8_MEMBER(cane_unknown_port0_w);
+	DECLARE_WRITE8_MEMBER(cane_76477_en_w);
+	DECLARE_WRITE8_MEMBER(cane_76477_dis_w);
```
Comment (cuavas) [inline src/mame/includes/8080bw.h]: Please don't add more stuff like this to the state class.  It's already too big, convoluted and poorly encapsulated.  Create a device to encapsulate the audio system for these games and instantiate it.  See the work I've done on mw8080bw.cpp this month for examples.

File: src/devices/sound/sn76477.cpp [hunk 5]
```diff
@@ -894,7 +894,9 @@ void sn76477_device::open_wav_file()
 {
 	char wav_file_name[30];
 
-	sprintf(wav_file_name, LOG_WAV_FILE_NAME, tag());
+	std::string s = tag();
+	std::replace(s.begin(), s.end(), ':', '_');	
+	sprintf(wav_file_name, LOG_WAV_FILE_NAME, s.c_str());
```
[the remark this answers is not in the document]
Comment (janniz) [reply src/devices/sound/sn76477.cpp]: Yes, it could be done that way, but I preferred to keep the most of the pre-existent code and to change only the parameter of the sprintf. 
Anyway if you prefer I can change also the other parts of the code.

File: src/devices/sound/sn76477.cpp (hunk 1 above)
Comment (janniz) [reply src/devices/sound/sn76477.cpp]: Indeed your minified expression is the same of that I've proposed but I find more human readable and obvious the long form.
I changed the condition because it was wrong. If LOG_WAV_ENABLED is defined as false the device will never write anything in the log wave file.

File: src/mame/includes/8080bw.h (hunk 4 above)
Comment (janniz) [reply src/mame/includes/8080bw.h]: I tried to follow the direction traced by the coders that added other games to that driver in the past 
(Eg:
	DECLARE_READ8_MEMBER(shuttlei_ff_r);
	DECLARE_WRITE8_MEMBER(shuttlei_ff_w);
	DECLARE_WRITE8_MEMBER(shuttlei_sh_port_1_w);
	DECLARE_WRITE8_MEMBER(shuttlei_sh_port_2_w);
)
but I could understant that code rules may change while the project evolve so I'm gonna recode this sound device.

**Comment by janniz** [comment]
> Quasi-related: Do you think you could enquire about this game and see if it had indeed been under development and if it had been given a title (perhaps "Pallanuoto"?)
> https://mamedev.emulab.it/undumped/index.php?title=Unknown_Water_Polo_Game
> 
> Also there is a rumor that Model Racing had licensed or perhaps cloned/bootlegged Looping, any word on that?

Pallanuoto was developed by Adolfo Melilli (the same coder of these two games), Giorgio Liverotti and Fausto Santi. It was still unreleased when Melilli and Liverotti left Model Racing and Adolfo doesn't know if it was finally released.

Same answer for Looping, he doesn't have any info about it, but for sure it was not licensed before he left.

File: src/devices/sound/sn76477.cpp (hunk 1 above)
Comment (cuavas) [reply src/devices/sound/sn76477.cpp]: Ah right, the condition makes sense.  It's generally better to avoid excessive levels of nesting and unnecessarily long lines.  now that I know what it's supposed to do I'd write it as `(LOG_WAV && (!m_enable || !LOG_WAV_ENABLED_ONLY))` – then it reads more naturally, like “log WAV and either enable is low or log when enabled only is unset“.

File: src/mame/includes/8080bw.h (hunk 4 above)
Comment (cuavas) [reply src/mame/includes/8080bw.h]: Yes, like many MAME drivers, this code was ported from C to C++ with minimal effort by just moving all the functions and static variables into one big class.  It hasn’t been cleaned up since, and more stuff has been hacked into it, making it even harder to understand and more effort to clean up.

The sound system for these new sets is fairly self-contained.  It should be easy to implement it as a device.

Comment (janniz) [reply src/mame/includes/8080bw.h]: I've never looked at MAME source code before and I've noticed that some other drivers have a more object oriented approach than others. When I've started coding that driver I was in doubt and lastly I choosed to implement the needed changes keeping the approach already present in the class.
Not a problem to implement it as a device. I'm going to do it as soon as I've some spare time this week. 

Thank you for your support.

File: src/devices/sound/sn76477.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (janniz) [reply src/devices/sound/sn76477.cpp]: Ok, I've used util::string_format as you suggested. Please look at my latest commit, thank you.

File: src/devices/sound/sn76477.cpp (hunk 1 above)
Comment (janniz) [reply src/devices/sound/sn76477.cpp]: Changed. Please look at my latest commit, thank you.

**Comment by stilett0** [comment]
> Pallanuoto was developed by Adolfo Melilli (the same coder of these two games), Giorgio Liverotti and Fausto Santi. It was still unreleased when Melilli and Liverotti left Model Racing and Adolfo doesn't know if it was finally released.
> 
> Same answer for Looping, he doesn't have any info about it, but for sure it was not licensed before he left.

Great, thanks!

File: src/mame/drivers/8080bw.cpp (hunk 3 above)
Comment (janniz) [reply src/mame/drivers/8080bw.cpp]: Moved...

File: src/mame/includes/8080bw.h (hunk 4 above)
Comment (janniz) [reply src/mame/includes/8080bw.h]: I refactored the "Cane" related code creating an audio device to encapsulate the audio system as requested.
I also refactored "Cane" and "Orbite" creating their own classes like you've done in mw8080bw

In order to move "Orbite" in its own class I had to move from private to protected some attributes of _8080bw_state (Eg: m_palette) and methods (Eg: clear_extra_columns and set_8_pixels)
I also had to add a new header file in audio/8080bw.h. Surprisingly the "8080-based black and white sound hardware" doesn't had a header file. For this file I didn't know how to format the head discalimer, so I might have done something in the wrong way, let me know if corrections are needed.

Lastly I've got a curiosity: I don't understand the need to have to different drivers for 8080bw and mw8080bw hardware. Why they can't be merged? They share the same hardware and architecture.

File: src/devices/sound/sn76477.cpp [hunk 6]
```diff
@@ -892,9 +892,11 @@ void sn76477_device::log_complete_state()
 
 void sn76477_device::open_wav_file()
 {
-	char wav_file_name[30];
+	std::string s = tag();
+	std::replace(s.begin(), s.end(), ':', '_');
+
+	char const* wav_file_name = util::string_format(LOG_WAV_FILE_NAME, s).c_str();
```
Comment (cuavas) [inline src/devices/sound/sn76477.cpp]: This creates a reference-to-temporary – the string created by `util::string_format` will fall out of scope at the end of the statement, so `wav_file_name` will point to deallocated memory.  You can make `wav_file_name` a `std::string` (`LOG` can use `std::string directly with no need to call `.c_str()` and you can call `.c_str()` for other funtions that need a C string).

File: src/mame/audio/8080bw.cpp [hunk 7]
```diff
@@ -1397,3 +1404,276 @@ WRITE8_MEMBER( _8080bw_state::darthvdr_08_w )
...
+*/
+DISCRETE_SOUND_END
+
+TIMER_DEVICE_CALLBACK_MEMBER(cane_audio_device::cane_vco_voltage_timer)
+{
+	double voltage;
+	voltage = 5 * (1 - exp(- (m_cane_vco_timer->fire_time().as_double() - m_cane_vco_rc_chargetime) / 47));
+
+	logerror("t = %d\n", m_cane_vco_timer->fire_time().as_double() - m_cane_vco_rc_chargetime);
+	logerror("vco_voltage = %d\n", voltage);
```
Comment (cuavas) [inline src/mame/audio/8080bw.cpp]: These should use `LOG` from `logmacro.h` so they don't spam unless `VERBOSE=1` is defined.

File: src/mame/audio/8080bw.h [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+class cane_audio_device : public device_t
+{
+public:
+	cane_audio_device(machine_config const &mconfig, char const *tag, device_t *owner, u32 clock = 0);
+
+	void cane_sh_port_1_w(u8 data);
+	void cane_music_w(u8 data);
+	void cane_76477_en_w(u8 data);
+	void cane_76477_dis_w(u8 data);
```
Comment (cuavas) [inline src/mame/audio/8080bw.h]: You don't need the `cane_` prefix on all of these - it's implied in the device name.

File: src/mame/audio/8080bw.h [hunk 9]
```diff
@@ -0,0 +1,44 @@
+// license:BSD-3-Clause
```
Comment (cuavas) [inline src/mame/audio/8080bw.h]: This file hasn't been added to the build scripts.

File: src/mame/drivers/8080bw.cpp [hunk 10]
```diff
@@ -3560,6 +3560,423 @@ void _8080bw_state::init_invmulti()
...
+	PORT_START("IN2")   // port 2
+	PORT_BIT( 0x02, IP_ACTIVE_HIGH, IPT_BUTTON1 )
+INPUT_PORTS_END
+
+
+void orbite_state::machine_start()
+{
+	m_scattered_colorram = std::make_unique<uint8_t []>(0x800);  
+	save_pointer(&m_scattered_colorram[0], "m_scattered_colorram", 0x800);
+	mw8080bw_state::machine_start();
```
Comment (cuavas) [inline src/mame/drivers/8080bw.cpp]: In general it's better to call the base `machine_start` first.

File: src/devices/sound/sn76477.cpp (hunk 6 above)
Comment (janniz) [reply src/devices/sound/sn76477.cpp]: My fault. I don't write C++ since years and I tought that .c_str() creates a new char* allocating the required space. Thx

File: src/mame/audio/8080bw.cpp (hunk 7 above)
Comment (janniz) [reply src/mame/audio/8080bw.cpp]: Ok

File: src/mame/audio/8080bw.h (hunk 9 above)
Comment (janniz) [reply src/mame/audio/8080bw.h]: Oh, sorry. I didn't take in account the build scripts. I compiled from command line with make and it compiled fine. 
I'm gonna look at the commit you've made in the source tree to understant what I missed.

File: src/mame/includes/8080bw.h (hunk 4 above)
Comment (cuavas) [reply src/mame/includes/8080bw.h]: The reasoning is that mw8080bw.cpp is for original Midway games, and 8080bw.cpp is for other games on similar hardware.  Ideally we'd be able to turn more common things into devices and not have as many dependencies between state classes.  When files or classes get too big, they become harder to understand.  A lot of MAME is only the way it is because we didn't have the modern framework in place when it was written, and it hasn't been updated yet.

File: src/mame/drivers/8080bw.cpp (hunk 10 above)
Comment (janniz) [reply src/mame/drivers/8080bw.cpp]: Thank you for your support. Sadly I've a lot to learn on mame internal working.

Comment (cuavas) [reply src/mame/drivers/8080bw.cpp]: Don't worry - MAME can seem big and unapproachable, but you can work your way up from small tasks to bigger tasks.  We're trying to make the core easier to work on and more approachable, too.

File: src/mame/audio/8080bw.h (hunk 9 above)
Comment (cuavas) [reply src/mame/audio/8080bw.h]: Yeah, it will build if you don't add a header to the scripts, but for people using features like building with Visual Studio the file won't appear in the IDE project, and dependencies may not be computed correctly.

File: src/mame/audio/8080bw.h (hunk 8 above)
Comment (cuavas) [reply src/mame/audio/8080bw.h]: I just noticed that `cane_76477_en_w` and `cane_76477_dis_w` aren't called from anywhere - is that intentional?
